### ISSUE TITLE
[inAppMessaging] Conflict with projects that use the latest version of picasso

### ISSUE THREAD

**Issue by khammami** [firebase/firebase-android-sdk#265]
### [READ] Step 1: Are you in the right place?

Issues filed here should be about bugs in __the code in this repository__. 
If you have a general question, need help debugging, or fall into some
other category use one of these other channels:

  * For general technical questions, post a question on [StackOverflow](http://stackoverflow.com/)
    with the firebase tag.
  * For general Firebase discussion, use the [firebase-talk](https://groups.google.com/forum/#!forum/firebase-talk)
    google group.
  * For help troubleshooting your application that does not fall under one
    of the above categories, reach out to the personalized
    [Firebase support channel](https://firebase.google.com/support/).

### [REQUIRED] Step 2: Describe your environment

  * Android Studio version: 3.3.2
  * Firebase Component: InAppMessaging display
  * Component version: 17.1.0

### [REQUIRED] Step 3: Describe the problem
I have the same problem as [this one](https://stackoverflow.com/questions/51933538/unable-to-use-firebase-in-app-messaging/51941220):
my app crash just after adding the dependency of inAppMessaging 

#### Steps to reproduce:

What happened? How can we make the problem occur?
Just use inAppMessaging with an app that already use picasso v2.71828

`
java.lang.NoClassDefFoundError: Failed resolution of: Lcom/squareup/picasso/OkHttpDownloader;
        at com.google.firebase.inappmessaging.display.internal.injection.modules.PicassoModule.providesFiamController(com.google.firebase:firebase-inappmessaging-display@@17.1.0:52)
        at com.google.firebase.inappmessaging.display.internal.injection.modules.PicassoModule_ProvidesFiamControllerFactory.get(com.google.firebase:firebase-inappmessaging-display@@17.1.0:34)
        at com.google.firebase.inappmessaging.display.internal.injection.modules.PicassoModule_ProvidesFiamControllerFactory.get(com.google.firebase:firebase-inappmessaging-display@@17.1.0:11)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.google.firebase.inappmessaging.display.internal.FiamImageLoader_Factory.get(com.google.firebase:firebase-inappmessaging-display@@17.1.0:21)
        at com.google.firebase.inappmessaging.display.internal.FiamImageLoader_Factory.get(com.google.firebase:firebase-inappmessaging-display@@17.1.0:8)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.google.firebase.inappmessaging.display.FirebaseInAppMessagingDisplay_Factory.get(com.google.firebase:firebase-inappmessaging-display@@17.1.0:62)
        at com.google.firebase.inappmessaging.display.FirebaseInAppMessagingDisplay_Factory.get(com.google.firebase:firebase-inappmessaging-display@@17.1.0:16)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.google.firebase.inappmessaging.display.internal.injection.components.DaggerAppComponent.providesFirebaseInAppMessagingUI(com.google.firebase:firebase-inappmessaging-display@@17.1.0:102)
        at com.google.firebase.inappmessaging.display.FirebaseInAppMessagingDisplayRegistrar.buildFirebaseInAppMessagingUI(com.google.firebase:firebase-inappmessaging-display@@17.1.0:74)
        at com.google.firebase.inappmessaging.display.FirebaseInAppMessagingDisplayRegistrar.access$lambda$0(com.google.firebase:firebase-inappmessaging-display@@17.1.0)
        at com.google.firebase.inappmessaging.display.FirebaseInAppMessagingDisplayRegistrar$$Lambda$1.create(com.google.firebase:firebase-inappmessaging-display@@17.1.0)
        at com.google.firebase.components.ComponentRuntime.lambda$new$0(com.google.firebase:firebase-common@@16.1.0:66)
        at com.google.firebase.components.ComponentRuntime$$Lambda$1.get(com.google.firebase:firebase-common@@16.1.0)
        at com.google.firebase.components.Lazy.get(com.google.firebase:firebase-common@@16.1.0:53)
        at com.google.firebase.components.ComponentRuntime.initializeEagerComponents(com.google.firebase:firebase-common@@16.1.0:155)
        at com.google.firebase.FirebaseApp.initializeAllApis(com.google.firebase:firebase-common@@16.1.0:779)
        at com.google.firebase.FirebaseApp.initializeApp(com.google.firebase:firebase-common@@16.1.0:369)
        at com.google.firebase.FirebaseApp.initializeApp(com.google.firebase:firebase-common@@16.1.0:332)
        at com.google.firebase.FirebaseApp.initializeApp(com.google.firebase:firebase-common@@16.1.0:316)
        at com.google.firebase.provider.FirebaseInitProvider.onCreate(com.google.firebase:firebase-common@@16.1.0:53)
        at android.content.ContentProvider.attachInfo(ContentProvider.java:1696)
        at android.content.ContentProvider.attachInfo(ContentProvider.java:1671)
        at com.google.firebase.provider.FirebaseInitProvider.attachInfo(com.google.firebase:firebase-common@@16.1.0:47)
        at android.app.ActivityThread.installProvider(ActivityThread.java:4999)
        at android.app.ActivityThread.installContentProviders(ActivityThread.java:4594)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4534)
        at android.app.ActivityThread.access$1500(ActivityThread.java:151)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1364)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:135)
        at android.app.ActivityThread.main(ActivityThread.java:5254)
        at java.lang.reflect.Method.invoke(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:372)
    	at com.android.internal.os.Zy
`

#### Relevant Code:

```
implementation ('com.squareup.picasso:picasso:2.71828',{
        exclude group: 'com.android.support'
    })
implementation 'com.google.firebase:firebase-inappmessaging-display:17.1.0'
```

**Comment by Neeraj0590** [comment]
I am facing this Problem 
Please help me to fix this

 FATAL EXCEPTION: main
    Process: com.bamberg, PID: 24399
    java.lang.NoClassDefFoundError: Failed resolution of: Lcom/squareup/picasso/OkHttpDownloader;
        at com.google.firebase.inappmessaging.display.internal.injection.modules.PicassoModule.providesFiamController(com.google.firebase:firebase-inappmessaging-display@@17.1.1:52)
        at com.google.firebase.inappmessaging.display.internal.injection.modules.PicassoModule_ProvidesFiamControllerFactory.get(com.google.firebase:firebase-inappmessaging-display@@17.1.1:34)
        at com.google.firebase.inappmessaging.display.internal.injection.modules.PicassoModule_ProvidesFiamControllerFactory.get(com.google.firebase:firebase-inappmessaging-display@@17.1.1:11)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.google.firebase.inappmessaging.display.internal.FiamImageLoader_Factory.get(com.google.firebase:firebase-inappmessaging-display@@17.1.1:21)
        at com.google.firebase.inappmessaging.display.internal.FiamImageLoader_Factory.get(com.google.firebase:firebase-inappmessaging-display@@17.1.1:8)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.google.firebase.inappmessaging.display.FirebaseInAppMessagingDisplay_Factory.get(com.google.firebase:firebase-inappmessaging-display@@17.1.1:62)
        at com.google.firebase.inappmessaging.display.FirebaseInAppMessagingDisplay_Factory.get(com.google.firebase:firebase-inappmessaging-display@@17.1.1:16)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.google.firebase.inappmessaging.display.internal.injection.components.DaggerAppComponent.providesFirebaseInAppMessagingUI(com.google.firebase:firebase-inappmessaging-display@@17.1.1:102)
        at com.google.firebase.inappmessaging.display.FirebaseInAppMessagingDisplayRegistrar.buildFirebaseInAppMessagingUI(com.google.firebase:firebase-inappmessaging-display@@17.1.1:74)
        at com.google.firebase.inappmessaging.display.FirebaseInAppMessagingDisplayRegistrar.access$lambda$0(Unknown Source:0)
        at com.google.firebase.inappmessaging.display.FirebaseInAppMessagingDisplayRegistrar$$Lambda$1.create(Unknown Source:2)
        at com.google.firebase.components.ComponentRuntime.lambda$new$0(com.google.firebase:firebase-common@@16.1.0:66)
        at com.google.firebase.components.ComponentRuntime$$Lambda$1.get(Unknown Source:4)
        at com.google.firebase.components.Lazy.get(com.google.firebase:firebase-common@@16.1.0:53)
        at com.google.firebase.components.ComponentRuntime.initializeEagerComponents(com.google.firebase:firebase-common@@16.1.0:155)
        at com.google.firebase.FirebaseApp.initializeAllApis(com.google.firebase:firebase-common@@16.1.0:779)
        at com.google.firebase.FirebaseApp.initializeApp(com.google.firebase:firebase-common@@16.1.0:369)
        at com.google.firebase.FirebaseApp.initializeApp(com.google.firebase:firebase-common@@16.1.0:332)
        at com.google.firebase.FirebaseApp.initializeApp(com.google.firebase:firebase-common@@16.1.0:316)
        at com.google.firebase.provider.FirebaseInitProvider.onCreate(com.google.firebase:firebase-common@@16.1.0:53)
        at android.content.ContentProvider.attachInfo(ContentProvider.java:1925)
        at android.content.ContentProvider.attachInfo(ContentProvider.java:1900)
        at com.google.firebase.provider.FirebaseInitProvider.attachInfo(com.google.firebase:firebase-common@@16.1.0:47)
        at android.app.ActivityThread.installProvider(ActivityThread.java:6643)
        at android.app.ActivityThread.installContentProviders(ActivityThread.java:6198)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6104)
        at android.app.ActivityThread.-wrap1(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1806)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:7036)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:441)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1408)
2019-03-26 17:54:32.987 24399-24399/com.bamberg E/AndroidRuntime: Caused by: java.lang.ClassNotFoundException: Didn't find class "com.squareup.picasso.OkHttpDownloader" on path: DexPathList[[zip file "/data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/base.apk", zip file "/data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_dependencies_apk.apk", zip file "/data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_resources_apk.apk", zip file "/data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_0_apk.apk", zip file "/data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_1_apk.apk", zip file "/data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_2_apk.apk", zip file "/data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_3_apk.apk", zip file "/data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_4_apk.apk", zip file "/data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_5_apk.apk", zip file "/data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_6_apk.apk", zip file "/data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_7_apk.apk", zip file "/data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_8_apk.apk", zip file "/data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_9_apk.apk"],nativeLibraryDirectories=[/data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/lib/arm64, /data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/base.apk!/lib/arm64-v8a, /data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_dependencies_apk.apk!/lib/arm64-v8a, /data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_resources_apk.apk!/lib/arm64-v8a, /data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_0_apk.apk!/lib/arm64-v8a, /data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_1_apk.apk!/lib/arm64-v8a, /data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_2_apk.apk!/lib/arm64-v8a, /data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_3_apk.apk!/lib/arm64-v8a, /data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_4_apk.apk!/lib/arm64-v8a, /data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_5_apk.apk!/lib/arm64-v8a, /data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_6_apk.apk!/lib/arm64-v8a, /data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_7_apk.apk!/lib/arm64-v8a, /data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_8_apk.apk!/lib/arm64-v8a, /data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_slice_9_apk.apk!/lib/arm64-v8a, /system/lib64, /vendor/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:125)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        	... 37 more
    	Suppressed: java.io.IOException: No original dex files found for dex location (arm64) /data/app/com.bamberg-U3GAh0twiE38Q2-1aw11bw==/split_lib_resources_apk.apk
        at dalvik.system.DexFile.openDexFileNative(Native Method)
        at dalvik.system.DexFile.openDexFile(DexFile.java:353)
        at dalvik.system.DexFile.<init>(DexFile.java:100)
        at dalvik.system.DexFile.<init>(DexFile.java:74)
        at dalvik.system.DexPathList.loadDexFile(DexPathList.java:374)
        at dalvik.system.DexPathList.makeDexElements(DexPathList.java:337)
        at dalvik.system.DexPathList.<init>(DexPathList.java:157)
        at dalvik.system.BaseDexClassLoader.<init>(BaseDexClassLoader.java:65)
        at dalvik.system.PathClassLoader.<init>(PathClassLoader.java:64)
        at com.android.internal.os.ClassLoaderFactory.createClassLoader(ClassLoaderFactory.java:73)
        at com.android.internal.os.ClassLoaderFactory.createClassLoader(ClassLoaderFactory.java:88)
        at android.app.ApplicationLoaders.getClassLoader(ApplicationLoaders.java:69)
        at android.app.ApplicationLoaders.getClassLoader(ApplicationLoaders.java:35)
        at android.app.LoadedApk.createOrUpdateClassLoaderLocked(LoadedApk.java:705)
        at android.app.LoadedApk.getClassLoader(LoadedApk.java:739)
        at android.app.LoadedApk.getResources(LoadedApk.java:980)
        at android.app.ContextImpl.createAppContext(ContextImpl.java:2420)
2019-03-26 17:54:32.987 24399-24399/com.bamberg E/AndroidRuntime:     at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6021)
        		... 8 more

**Comment by shahzaib-ch** [comment]
I am also facing same issue as @Neeraj0590

**Comment by Neeraj0590** [comment]
hi
you need to remove Picasso lib from your other dependency then this error
is resolved

Regards
Neeraj singh

**Comment by shahzaib-ch** [comment]
@Neeraj0590 solved it using dependency resolution. thanks

**Comment by audkar** [comment]
This is not good if Firebase library is preventing app to use latest version of other libraries :(

**Comment by chaitukadivella** [comment]
I am also facing same issue. I cant remove Picasso dependency as it is extensively used in my app. as per suggestion in SO, Picasso.get() has to be replaced with Picasso.with(context). Changing like this is a lot of work for me. It would be great if firebase in-app messaging has an upgraded version of Picasso.

**Comment by safety97** [comment]
Exactly the same problem here!
 Firebase can you update your library, please? I really need to use this feature in my app, but I'm stuck with the library conflicts.

**Comment by yong0726** [comment]
Same issue here!. Could you please increase the priority of this?

**Comment by way2jatin** [comment]
Firebase InAppMessaging has been recently updated to 18.0.0. Can anyone please confirm whether this issue still remains?

**Comment by khammami** [comment]
@way2jatin still the same with v18.0.1

To fix it, you must downgrade Picasso version to 2.5.2, or remove it.

**Comment by way2jatin** [comment]
Thanks @khammami for the info. I wonder when will they fix this important issue.

**Comment by prsolucoes** [comment]
Please Firebase Team upgrade some libraries to latest version:

```
implementation 'com.squareup.okhttp3:okhttp:3.14.2'
implementation 'com.squareup.picasso:picasso:2.71828'
```

Thanks.

**Comment by achuinard** [comment]
Was in the process of making an issue when I came across this. I hope the team can upgrade Picasso in the inappmessaging-display Gradle ASAP. I found there were performance improvements and bugfixes in the latest version which is why I moved off of 2.5.2 in the first place. There was even a case where an image wouldn't load in 2.5.2 and did load with the latest version, but I have to go backwards now.
